File: packages/shared/src/connectors/coinmarketcal.ts
import { fetchJson } from "./base.js";

// apiKey is an explicit parameter, not read from a shared config module --
// this file is imported by both the Node backend (its own config.ts reads
// COINMARKETCAL_API_KEY from process.env) and the browser frontend/mobile
// app (their own local settings storage), neither of which the other can see.

const CMC_CAL = "https://api.coinmarketcal.com/v2";

export interface CoinMarketCalCoin {
  slug: string;
  symbol: string;
  name: string;
}

// Confirmed against a live response (2026-07-26) -- CoinMarketCal's docs
// 403'd while this was first built, so this shape replaces an earlier
// best-guess version that had two wrong field names (date_event -> date,
// source -> sourceUrl). No `categories` field is present on real events
// despite being documented in third-party client libraries; event-type
// classification falls back to the title alone.
export interface CoinMarketCalEvent {
  id: string;
  slug: string;
  title: string;
  description: string | null;
  date: string;
  dateEnd?: string;
  coins: CoinMarketCalCoin[];
  categories?: { name: string }[];
  sourceUrl: string | null;
}

function toDateRangeParam(date: Date): string {
  const mm = String(date.getMonth() + 1).padStart(2, "0");
  const dd = String(date.getDate()).padStart(2, "0");
  return `${mm}/${dd}/${date.getFullYear()}`;
}

/**
 * Fetches upcoming events across ALL coins (not filtered server-side by our
 * watchlist) for a date window, and lets the caller match events to our
 * tokens itself. CoinMarketCal's own coin identifiers aren't guaranteed to
 * line up with our coingeckoId values, so matching client-side by
 * ticker/name -- with the same collision caution used everywhere else in
 * this app -- is safer than trusting a server-side `coins=` filter to
 * resolve the right asset.
 */
export async function fetchCoinMarketCalEvents(
  apiKey: string | undefined,
  daysAhead = 90,
  max = 300
): Promise<CoinMarketCalEvent[] | null> {
  if (!apiKey) return null;

  const now = new Date();
  const until = new Date(now.getTime() + daysAhead * 24 * 60 * 60 * 1000);

  const params = new URLSearchParams({
    dateRangeStart: toDateRangeParam(now),
    dateRangeEnd: toDateRangeParam(until),
    max: String(max),
    sortBy: "created_desc",
  });

  const data = await fetchJson<{ data?: CoinMarketCalEvent[] }>(`${CMC_CAL}/events?${params}`, {
    headers: { "x-api-key": apiKey },
  });
  return data?.data ?? null;
}

export function eventTitle(event: CoinMarketCalEvent): string {
  return event.title || "Untitled event";
}

File: mobile/lib/connectors/coinmarketcal.dart
// Mirrors packages/shared/src/connectors/coinmarketcal.ts -- see that file's
// own comment for the collision-caution reasoning behind why matching is
// left entirely to the caller (this connector just fetches the raw feed).
import 'http_client.dart';

const _cmcCal = 'https://api.coinmarketcal.com/v2';

class CoinMarketCalCoin {
  final String symbol;
  final String name;
  CoinMarketCalCoin({required this.symbol, required this.name});
}

class CoinMarketCalEvent {
  final String title;
  final String date;
  final List<CoinMarketCalCoin> coins;
  final List<String> categoryNames;
  final String? sourceUrl;

  CoinMarketCalEvent({
    required this.title,
    required this.date,
    required this.coins,
    required this.categoryNames,
    this.sourceUrl,
  });

  String get eventTitle => title.isNotEmpty ? title : 'Untitled event';
}

String _dateRangeParam(DateTime d) {
  final mm = d.month.toString().padLeft(2, '0');
  final dd = d.day.toString().padLeft(2, '0');
  return '$mm/$dd/${d.year}';
}

Future<List<CoinMarketCalEvent>?> fetchCoinMarketCalEvents(
  String? apiKey, {
  int daysAhead = 90,
  int max = 300,
}) async {
  if (apiKey == null || apiKey.isEmpty) return null;

  final now = DateTime.now();
  final until = now.add(Duration(days: daysAhead));

  final params = {
    'dateRangeStart': _dateRangeParam(now),
    'dateRangeEnd': _dateRangeParam(until),
    'max': '$max',
    'sortBy': 'created_desc',
  };
  final uri = Uri.parse('$_cmcCal/events').replace(queryParameters: params);

  final data = await fetchJson(uri.toString(), headers: {'x-api-key': apiKey});
  if (data == null) return null;
  final events = (data['data'] as List?) ?? [];
  return events.map((e) {
    final json = e as Map<String, dynamic>;
    final coins = (json['coins'] as List? ?? [])
        .map((c) => CoinMarketCalCoin(symbol: (c['symbol'] as String? ?? '').toUpperCase(), name: c['name'] as String? ?? ''))
        .toList();
    final categories = (json['categories'] as List? ?? [])
        .map((c) => ((c as Map<String, dynamic>)['name'] as String? ?? '').toLowerCase())
        .toList();
    return CoinMarketCalEvent(
      title: json['title'] as String? ?? '',
      date: json['date'] as String? ?? '',
      coins: coins,
      categoryNames: categories.cast<String>(),
      sourceUrl: json['sourceUrl'] as String?,
    );
  }).toList();
}
